src/doc/trpl/closures.md

```diff
@@ -1,214 +1,510 @@
+us a large amount of control over what kind of closures we can take.
+
+The `|| {}` syntax for closures is sugar for these three traits. For example,
+when you create a closure like this:
 
-```{rust}
-let square = |x: i32| { x * x };
+```rust
+let num = 5;
+let add_num = |x| x + num;
+
+add_num(1);
 ```
 
-We've seen this before. We make a closure that takes an integer, and returns
-its square.
+Rust desugars that code to this code:
 
-```{rust}
-# fn twice<F: Fn(i32) -> i32>(x: i32, f: F) -> i32 { f(x) + f(x) }
-# let square = |x: i32| { x * x };
-twice(5, square); // evaluates to 50
 ```
+# #![feature(unboxed_closures)]
 
-This line is more interesting. Here, we call our function, `twice`, and we pass
-it two arguments: an integer, `5`, and our closure, `square`. This is just like
-passing any other two variable bindings to a function, but if you've never
-worked with closures before, it can seem a little complex. Just think: "I'm
-passing two variables: one is an i32, and one is a function."
+struct ClosureEnvironment<'env> {
+    num: &'env i32
+}
 
-Next, let's look at how `twice` is defined:
+impl<'env> Fn<(i32,)> for ClosureEnvironment<'env> {
+    type Output = i32;
 
-```{rust,ignore}
-fn twice<F: Fn(i32) -> i32>(x: i32, f: F) -> i32 {
+    extern "rust-call" fn call(&self, (x,): (i32,)) -> i32 {
+        x + *self.num
+    }
+}
+# fn main() {
+let num = 5;
+let env = ClosureEnvironment { num: &num };
+
+env.call((1,));
+# }
 ```
 
-`twice` takes two arguments, `x` and `f`. That's why we called it with two
-arguments. `x` is an `i32`, we've done that a ton of times. `f` is a function,
-though, and that function takes an `i32` and returns an `i32`. This is
-what the requirement `Fn(i32) -> i32` for the type parameter `F` says.
-Now `F` represents *any* function that takes an `i32` and returns an `i32`.
+Whew! We create a new struct, `ClosureEnvironment`, and then implement `Fn` for
+it. Instead of the `|| {}` syntax, we make a new `ClosureEnvironment` and then
+call `call()` on it.
+
+Rust will generate a new `struct` and implement the correct `Fn`, `FnMut`, or
+`FnOnce` trait for each closure you write. This means the name won't literally
+be `ClosureEnvironment`, as it's unique per closure.
 
-This is the most complicated function signature we've seen yet! Give it a read
-a few times until you can see how it works. It takes a teeny bit of practice, and
-then it's easy. The good news is that this kind of passing a closure around
-can be very efficient. With all the type information available at compile-time
-the compiler can do wonders.
+# Taking closures as arguments
 
-Finally, `twice` returns an `i32` as well.
+Now that we know that closures are traits, we already know how to accept and
+return closures: just like any other trait!
 
-Okay, let's look at the body of `twice`:
+This also means that we can choose static vs dynamic dispatch as well. First,
+let's write a function which takes a closure, calls it, and returns the
```

Comment:
Technically wrong -- `Fn*` are callables and aren't necessarily closures.

"First, let's write a function which takes a callable, calls it, and returns the result, and then pass a closure to it." might be clearer.